### PR TITLE
Article feature

### PR DESCRIPTION
**Pull request by AKidd95**
#### Description
Changes proposed in this pull request (we went a little out of scope on this one)

* create a form for adding articles
* wrote acceptance and unit tests for article
#### What I have learned working on this feature: 
* how to properly indent with haml
* writing forms
* proper routing 
* reinforce controller action creation knowledge

schermata 2018-03-15 alle 17 45 23
schermata 2018-03-15 alle 17 45 36
schermata 2018-03-15 alle 17 46 40

### CODE REVIEW

File: features/create_articles.feature [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Feature: Create articles
+  As a journalist
+  in order to attract subscribers
```
Comment (davidboll) [inline features/create_articles.feature]: Please change to large I.. "In order.."

File: features/create_articles.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  I would like to be able to create articles
+
+  Background:
+    Given I visit the "index" page
+    When I click "New Article" link
+
```
Comment (davidboll) [inline features/create_articles.feature]: Unnessesary use of Background, move these steps to Scenario. And change "When " to "And"

**Review by davidboll — CHANGES_REQUESTED** [review]
See comments

**Comment by davidboll** [comment]
Please remove Channels, Jobs and Mailers if they arn't being used. See https://coveralls.io/builds/15997151

**Review by davidboll — APPROVED** [review]

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
```
Comment (isabellelii) [inline config/routes.rb]: Remove these comments

**Review by isabellelii — APPROVED** [review]

**Review by kmthorsnes — APPROVED** [review]
Looks good to me

**Comment by tochman** [comment]
Please stick to one PR per feature. If you feel that the scope is to narrow on a feature, please inform your team and bring it up on that next scrum. But generally, make sure you stay in scope.

File: .gitignore [hunk 4]
```diff
@@ -27,3 +27,4 @@
 /config/master.key
 
 # Ignore others
+/.idea/
```
Comment (tochman) [inline .gitignore]: This won't ignore the `.idea` files. Please remove the first `/`

File: app/channels/application_cable/channel.rb [hunk 5]
```diff
@@ -1,4 +0,0 @@
-module ApplicationCable
-  class Channel < ActionCable::Channel::Base
-  end
-end
```
Comment (tochman) [inline app/channels/application_cable/channel.rb]: Please stay in scope. This file was removed in another PR by another developer (that should also have stayed in scope). Potential merge conflict here...

File: app/channels/application_cable/connection.rb [hunk 6]
```diff
@@ -1,4 +0,0 @@
-module ApplicationCable
-  class Connection < ActionCable::Connection::Base
-  end
-end
```
Comment (tochman) [inline app/channels/application_cable/connection.rb]: Same here...

File: app/mailers/application_mailer.rb [hunk 7]
```diff
@@ -1,4 +0,0 @@
-class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
-  layout 'mailer'
-end
```
Comment (tochman) [inline app/mailers/application_mailer.rb]: Removing mailer when you use Devise might not be the best thing. Not a blocker but be prepared to add this back in when making use of some of the devise functionality (like sending emails to new users for confirmation, etc)

File: app/models/article.rb [hunk 8]
```diff
@@ -0,0 +1,2 @@
+class Article < ApplicationRecord
+end
```
Comment (tochman) [inline app/models/article.rb]: No validations? Are you happy with users creating articles without a `title` or a `content`?

File: app/controllers/articles_controller.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
+class ArticlesController < ApplicationController
+  def index
+    @article = Article.all
+  end
+
+  def new
+    @article = Article.new
```
Comment (tochman) [inline app/controllers/articles_controller.rb]: I see that you use:
``` 
form_with model: Article.new ...
```
on your view. In this case, you don't need this variable. I would keep it though, and change my form to 
```
form_with model: @article
```

File: app/controllers/articles_controller.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def create
+    @article = Article.new(article_params)
+    if @article.save
+      redirect_to @article, notice: "Article was successfully created."
+    else
+      render 'new'
```
Comment (tochman) [inline app/controllers/articles_controller.rb]: I would flash the error messages if the article fails to save.

File: app/views/articles/new.html.haml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+= form_with model: Article.new, url: articles_path, local: true do |form|
```
Comment (tochman) [inline app/views/articles/new.html.haml]: Please see my comment on the `ArticlesController`. Also, when you use the `model:` setting, you can probably omit the `url: `

Comment (tochman) [reply app/views/articles/new.html.haml]: +1 for remembering `local: true`

File: app/views/articles/new.html.haml [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  %p
+    = form.label :title
+    %br/
+    = form.text_field :title, id: :article_title
+  %p
+    = form.label :content
+    %br/
+    = form.text_area :content, id: :article_content
+  %p
+    = form.submit placeholder: "Create Article"
```
Comment (tochman) [inline app/views/articles/new.html.haml]: Not sure if you should use `placeholder:` here.  `form_with` is relatively new to me, but I usually use `value:`

**Review by tochman — CHANGES_REQUESTED** [review]

File: app/views/articles/new.html.haml (hunk 12 above)
Comment (AKidd95) [reply app/views/articles/new.html.haml]: They both work but I will change to value for fear of the unknown

File: app/mailers/application_mailer.rb (hunk 7 above)
Comment (AKidd95) [reply app/mailers/application_mailer.rb]: will keep this in mind for the future. We will try to make more a conscious effort to stay in scope 👍

File: features/step_definitions/create_articles_steps.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+Given("I visit the {string} page") do |string|
+  visit root_path
```
Comment (oliverochman) [inline features/step_definitions/create_articles_steps.rb]: You don't use the variable here

File: app/controllers/articles_controller.rb [hunk 14]
```diff
@@ -0,0 +1,19 @@
+class ArticlesController < ApplicationController
+  def index
+    @article = Article.all
```
Comment (oliverochman) [inline app/controllers/articles_controller.rb]: When is this path actually accessed? Am I missing something?

File: app/views/articles/index.html.haml [hunk 15]
```diff
@@ -0,0 +1 @@
+%h1 Welcome to PAPs (Passive Aggressive Publishers)
```
Comment (oliverochman) [inline app/views/articles/index.html.haml]: Why is this view here? And if it is needed why aren't you showing the articles?
IMO it feels out of scope

File: features/create_articles.feature [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+  I would like to be able to create articles
+
+  Background:
+    Given I visit the "index" page
+
+  Scenario: Successfully create an article
+    When I click "New Article" link
+    Then I fill in "Title" with "A Whole New World"
+    And I fill in "Content" with "A new fantastic point of view"
+    And I click "Create Article" button
```
Comment (oliverochman) [inline features/create_articles.feature:13]: And then what?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: features/create_articles.feature (hunk 16 above)
Comment (AKidd95) [reply features/create_articles.feature:13]: yeah we added a Then test that checks for the user to visit the homepage after successful creation

File: features/step_definitions/create_articles_steps.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+When("I fill in {string} with {string}") do |input, value|
+  fill_in(input, with: value)
+end
+
+When("I click {string} button") do |string|
+  click_button string
+end
+
+Then("I should be redirected to Homepage") do
+
```
Comment (oliverochman) [inline features/step_definitions/create_articles_steps.rb]: This is empty

**Review by oliverochman — APPROVED** [review]
Looks good to me!!!

File: features/step_definitions/create_articles_steps.rb [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+When("I click {string} button") do |string|
+  click_button string
+end
+
+Then("I should be redirected to Homepage") do
+  expect(current_path).to eq root_path
+end
+
+Then("I should stay on new_article_path") do
+  expect(current_path).to eq current_path
```
Comment (tochman) [inline features/step_definitions/create_articles_steps.rb]: That does NOT mean much. Sorry, but no.

`Then the color of my pans should be what the color of my pants is`

File: features/create_articles.feature [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    Then I should be redirected to Homepage
+
+  Scenario: Failed to fill in content field
+    When I click "New Article" link
+    Then I fill in "Title" with "A Whole New World"
+    Then I fill in "Content" with " "
+    And I click "Create Article" button
+    Then I should stay on new_article_path
```
Comment (tochman) [inline features/create_articles.feature]: Please avoid using snake case in tests/scenarios

File: spec/models/article_spec.rb [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+
+RSpec.describe Article, type: :model do
+  describe 'DB columns' do
+    it { is_expected.to have_db_column :title }
+    it { is_expected.to have_db_column :content }
+  end
+
+  describe 'Factory' do
+    it 'should have valid Factory' do
+      expect(FactoryBot.create(:article)).to be_valid
```
Comment (tochman) [inline spec/models/article_spec.rb]: Please use the shorter `create(:article)`

**Review by tochman — CHANGES_REQUESTED** [review]
Not quite there yet.

File: spec/models/article_spec.rb (hunk 20 above)
Comment (AKidd95) [reply spec/models/article_spec.rb]: I am not quite sure what this means, Ive looked through some past code of ours that uses factory bot and they all test the same way. I don't see the issue

Comment (tochman) [reply spec/models/article_spec.rb]: I’ve demonstrated several times how we can include FactoryBot syntax methods to your rails_helper and use this shortened syntax in your specs.

**Review by tochman — APPROVED** [review]
